enterprise/server/remote_execution/copy_on_write/copy_on_write.go

```diff
@@ -860,43 +877,60 @@ func mmapDataFromFd(fd, size int) ([]byte, error) {
 	return data, nil
 }
 
-func (m *Mmap) initMap() error {
-	m.mu.Lock()
-	defer m.mu.Unlock()
+func (m *Mmap) initMap() (err error) {
+	m.initMu.Lock()
+	defer m.initMu.Unlock()
 	if m.closed {
 		return status.InternalError("store is closed")
 	}
-	if m.source != snaputil.ChunkSourceUnmapped {
+	if m.data != nil {
 		return nil
 	}
-	if m.lazyDigest == nil {
-		return status.InternalError("cannot initialize chunk without a digest")
+	path := m.path()
+	if m.lazyDigest != nil {
+		if err := m.fetch(path); err != nil {
+			return err
+		}
 	}
-
-	outputPath := filepath.Join(m.dataDir, ChunkName(m.Offset, false /*dirty*/))
-	artifactSrc, err := snaputil.GetArtifact(m.ctx, m.env.GetFileCache(), m.env.GetByteStreamClient(), m.remoteEnabled, m.lazyDigest, m.remoteInstanceName, outputPath)
+	data, err := mmapDataFromPath(path)
 	if err != nil {
-		return status.WrapErrorf(err, "fetch snapshot chunk for offset %d digest %s", m.Offset, m.lazyDigest.Hash)
+		return status.WrapErrorf(err, "create mmap for path %s", path)
 	}
+	m.data = data
+
+	return nil
+}
+
+func (m *Mmap) path() string {
+	return filepath.Join(m.dataDir, ChunkName(m.Offset, m.dirty))
+}
 
-	data, err := mmapDataFromPath(outputPath)
+// Fetch fetches the chunk from cache if applicable.
+func (m *Mmap) Fetch() error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.fetch(m.path())
+}
+
+func (m *Mmap) fetch(path string) error {
+	if m.fetched || m.lazyDigest == nil {
```

Comment:
Should we return an error if m.fetched == false && m.lazyDigest == nil? Otherwise in mmapDataFromPath we'll try to open a file that doesn't exist

Reply:
I think it's fine to just have mmapDataFromPath return a NotExist error.

Reply:
From a debugging standpoint it might be helpful to know if the lazyDigest is not set, because that should never be the case and would indicate a bug. Don't feel that strongly though

Reply:
Fair, done. Had to set `fetched: true` in the non-lazy constructors for this to work.